lib/honeybadger/http_adapter.ex

```diff
@@ -0,0 +1,173 @@
+defmodule Honeybadger.HTTPAdapter do
+  @moduledoc """
+  HTTP adapter helper module.
+
+  You can configure the HTTP adapter that Honeybadger uses by setting the
+  following option:
+
+      http_adapter: Honeybadger.HTTPAdapter.Req
+
+  Default options can be set by passing a list of options:
+
+      http_adapter: {Honeybadger.HTTPAdapter.Req, [...]}
+
+  You can also set the client for the application:
+
+      config :honeybadger, :http_adapter, Honeybadger.HTTPAdapter.Req
+
+  ## Usage
+      defmodule MyApp.MyHTTPAdapter do
+        @behaviour Honeybadger.HTTPAdapter
+
+        @impl true
+        def request(method, url, body, headers, opts) do
+          # ...
+        end
+      end
+  """
+
+  alias Honeybadger.{InvalidResponseError, ServerUnreachableError}
+
+  defmodule HTTPResponse do
+    @moduledoc """
+    Struct used by HTTP adapters to normalize HTTP responses.
+    """
+
+    @type header :: {binary(), binary()}
+    @type t :: %__MODULE__{
+            http_adapter: atom(),
+            request_url: binary(),
+            status: integer(),
+            headers: [header()],
+            body: binary() | term()
+          }
+
+    defstruct http_adapter: nil, request_url: nil, status: 200, headers: [], body: ""
+
+    def format(response) do
+      [request_url | _rest] = String.split(response.request_url, "?", parts: 2)
+
+      """
+      HTTP Adapter: #{inspect(response.http_adapter)}
+      Request URL: #{request_url}
+
+      Response status: #{response.status}
+
+      Response headers:
+      #{Enum.reduce(response.headers, "", fn {k, v}, acc -> acc <> "\n#{k}: #{v}" end)}
+
+      Response body:
+      #{inspect(response.body)}
+      """
+    end
+  end
+
+  @type method :: :get | :post
+  @type body :: binary() | nil
+  @type headers :: [{binary(), binary()}]
+
+  @callback request(method(), binary(), body(), headers(), Keyword.t()) ::
+              {:ok, map()} | {:error, any()}
+
+  @doc """
+  Sets a user agent header.
+
+  The header value will be `Honeybadger-VERSION` with VERSION being the `:vsn` of
+  `:honeybadger` app.
+  """
+  @spec user_agent_header() :: {binary(), binary()}
+  def user_agent_header do
+    version = Application.spec(:honeybadger, :vsn) || "0.0.0"
```

Comment:
The application has to be loaded before this call is made, so I think it can safely omit the `||`:

```suggestion
    version = Application.spec(:honeybadger, :vsn)
```